src/Pos/Block/Network/Logic.hs

```diff
@@ -150,20 +152,34 @@ requestTip _ nodeId conv = do
 -- Headers processing
 ----------------------------------------------------------------------------
 
+-- | Result of creating a request message for more headers.
+data MkHeadersRequestResult
+    = MhrrBlockAdopted
+      -- ^ The block pointed by the header is already adopted, no need to
+      -- make the request.
+    | MhrrWithoutCheckpoints MsgGetHeaders
+      -- ^ The request can be made, but no checkpoints available.
+    | MhrrWithCheckpoints MsgGetHeaders
+      -- ^ A good request with checkpoints can be made.
+
 -- | Make 'GetHeaders' message using our main chain. This function
 -- chooses appropriate 'from' hashes and puts them into 'GetHeaders'
 -- message.
 mkHeadersRequest
     :: forall ssc m.
        WorkMode ssc m
-    => HeaderHash -> m (Maybe MsgGetHeaders)
+    => HeaderHash -> m MkHeadersRequestResult
 mkHeadersRequest upto = do
     uHdr <- blkGetHeader @ssc upto
-    runMaybeT $ do
-        -- no reason to make a request when we already have the latest header
-        guard (isNothing uHdr)
-        bHeaders <- MaybeT $ nonEmpty . toList <$> getHeadersOlderExp @ssc Nothing
-        pure $ MsgGetHeaders (toList bHeaders) (Just upto)
+    if isJust uHdr then return MhrrBlockAdopted else do
+        mCheckpoints <- nonEmpty . toList <$> getHeadersOlderExp @ssc Nothing
+        case mCheckpoints of
+            Nothing -> do
```

Comment:
Not sure about this codepath. Can it ever happen? When does `getHeadersOlderExp` return an empty list? @volhovM